sbin/tpm2-evt-log-parser.awk

```diff
@@ -0,0 +1,135 @@
+#!/usr/bin/gawk -bf
+@load "readfile"
+
+function assert(condition, string)
+{
+	if (!condition) {
+		print string
+		exit 1
+	}
+}
+
+function ord_init()
+{
+	for (_i = 0; _i < 256; _i++) {
+		ord[sprintf("%c", _i)] = _i
+	}
+}
+
+function x2n(hex, width)
+{
+	mult = 1
+	num = 0
+	for (_i = 0; _i < width; _i++) {
+		num += ord[substr(hex, _i+1, 1)] * mult
+		mult *= 256
+	}
+	return num
+}
+
+function hexdump(hex, len)
+{
+	for (_i = 0; _i < len; _i++) {
+		printf("%02x", ord[substr(hex, _i+1, 1)])
+	}
+}
+
+function alg_name(id)
+{
+	switch (id) {
+		case 0x0004: return "SHA1"
+		case 0x000b: return "SHA256"
+		case 0x000c: return "SHA384"
+		case 0x000d: return "SHA512"
+		case 0x0012: return "SM3-256"
+		case 0x0027: return "SHA3-256"
+		case 0x0028: return "SHA3-384"
+		case 0x0029: return "SHA3-512"
+		default: return sprintf("unknown (%#06)", id)
+	}
+}
+
+function string_or_hex(str, len)
+{
+	_len = len
+	if (_len > 300)
+		_len = 128
+	# String must start with a series of printable characters ...
+	if (match(str, "[[:graph:][:blank:]]*", a) != 1) {
+		hexdump(str, _len)
+	# ... long until the end, with "optional" (i.e. bad implementation) \0.
+	} else if (len != a[0, "length"] &&
+		   (len != a[0, "length"] + 1 || index(str, "\0") != len)) {
+		hexdump(str, _len)
+	} else
+		printf("%.*s", _len, a[0])
+	if (_len != len)
+		printf("... (event truncated to %d first bytes)", _len)
```

Comment:
Maybe this should print untruncated size as well.

Reply:
https://github.com/TrenchBoot/qubes-antievilmaid/compare/83d64223fc8dedf1ddbf8583ef1a37ebe1689715..2ccd86f8c75b9c905ff792c1b06f3de7bbf96361#diff-b56cabcce2405fa62574044a34233aab91df4bcc597a3633575f4e551484a7c7R67

https://github.com/TrenchBoot/qubes-antievilmaid/compare/83d64223fc8dedf1ddbf8583ef1a37ebe1689715..2ccd86f8c75b9c905ff792c1b06f3de7bbf96361#diff-22a962ec8ef35e178e5425797ae4760edd394427c0dd4b1c8a94ccf619cc219aR52